app/src/main/java/com/adam/aslfms/service/applemusic/TrackData.java

```diff
@@ -0,0 +1,145 @@
+package com.adam.aslfms.service.applemusic;
+
+import java.util.ArrayList;
+import java.util.Date;
+
+/**
+ * Created by 4-Eyes on 16/3/2017.
+ *
+ */
+
+class TrackData {
+
+    private String artist;
+    private String title;
+    private String album;
+    private Date startTime;
+    private PlayingState currentState = PlayingState.UNKNOWN;
+    private ArrayList<Long> playTimes = new ArrayList<>();
+    private long lastStateChangedTime;
+    private boolean isRepeat = false;
+
+    void setContentType(String contentType) {
+        switch (contentType) {
+            case "Pause":
+                currentState = PlayingState.PLAYING;
+                break;
+            case "Play":
+                currentState = PlayingState.PAUSED;
+                break;
+            default:
+                currentState = PlayingState.UNKNOWN;
+                break;
+        }
+        this.lastStateChangedTime = System.currentTimeMillis();
+    }
+
+    boolean mergeSame(TrackData data) {
+        if (this.currentState.equals(data.currentState)) return false;
+        if (this.currentState.equals(PlayingState.PLAYING)
+                && data.currentState.equals(PlayingState.PAUSED)) {
+            playTimes.add(data.startTime.getTime() - this.lastStateChangedTime);
+        }
+        this.currentState = data.currentState;
+        lastStateChangedTime = System.currentTimeMillis();
+        return true;
+    }
+
+    long recordedTotalPlayTime() {
+        long total = 0;
+        for (long playtime : playTimes) {
+            total += playtime;
+        }
+        return total;
+    }
+
+    long currentTotalPlayTime() {
+        long recordedTime = this.recordedTotalPlayTime();
+        if (currentState.equals(PlayingState.PLAYING)) {
+            recordedTime += System.currentTimeMillis() - lastStateChangedTime;
+        }
+        return recordedTime;
+    }
+
+    long finalisePlayTime(long currentTrackDuration) {
+        if (currentState.equals(PlayingState.PLAYING)) {
+            long lastPlayTime = System.currentTimeMillis() - lastStateChangedTime;
+            long totalPlayTimes = recordedTotalPlayTime();
+            long overlapTime = (totalPlayTimes + lastPlayTime - currentTrackDuration);
+            if (totalPlayTimes + lastPlayTime > currentTrackDuration + 5000 &&
+                    overlapTime < currentTrackDuration) { // check overlap time is not ridiculous
+                playTimes.add(lastPlayTime - overlapTime);
+                return overlapTime;
+            }
+            playTimes.add(lastPlayTime);
+        }
+        return 0;
+    }
+
+    boolean sameTrack(TrackData other) {
+        return this.title != null && this.artist != null && this.album != null &&
+                other.title != null && other.artist != null && other.album != null &&
```

Comment:
Do you need to check whether the fields on other aren't null? "some_string".equals(null) should be fine, right?